### PR TITLE
DDP refactoring: Extract parameter layout computation into optimizer classmethod

### PR DESCRIPTION
**Pull request by deepakn94**
At a high level, this PR refactors `DistributedDataParallel` (DDP). DDP's constructor now accepts a `full_param_layout` argument that describes how parameters and gradients should be mapped in the underlying `_ParamAndGradBuffer`. Distributed optimizers compute this mapping via a static `compute_full_param_layout` method. This allows different optimizers to control gradient buffer layout — e.g., the existing `DistributedOptimizer` uses a sequential layout with padding to ensure buckets are multiples of the data-parallel size, and in the future, `LayerwiseOptimizer` for Muon could specify a mapping such that each gradient bucket shard contains a full layer's gradients. This ensures that: a) gradient reduction can use reduce-scatter instead of all-reduce (gradient and parameter sharding follows optimizer sharding exactly), and b) no extra parameter copies are needed after the parameter all-gather (results go directly into the right place for the next forward pass).

**Key changes:**
- New `param_layout.py` module with `BufferKey`, `PerBufferParamLayout`, and `FullParamLayout` dataclasses, plus shared padding utilities (`pad_param_start`, `pad_bucket_end`).
- `DistributedOptimizer.compute_full_param_layout()` and `_compute_per_buffer_param_layout()` static methods that pre-compute padded layouts.
- `DistributedDataParallel` accepts an optional `full_param_layout` and passes per-buffer layouts to `_ParamAndGradBuffer`.
- `_ParamAndGradBuffer` consumes pre-computed layouts when provided, falling back to `_compute_default_per_buffer_param_layout()` (no-padding layout for non-distributed-optimizer cases).
- `training.py` pre-computes layouts via `DistributedOptimizer.compute_full_param_layout()` when using the distributed optimizer with DDP.
- New unit tests in `test_param_layout.py` for layout computation functions.

In the future, `training.py` can assemble layouts from multiple distributed optimizers (this should be straightforward since each optimizer will have disjoint buffers).

### CODE REVIEW

File: megatron/core/distributed/param_and_grad_buffer.py [hunk 1]
```diff
@@ -806,103 +889,15 @@ def _pad_end_of_bucket_if_needed(bucket_end_index: int) -> int:
             Pads end index of bucket if using distributed optimizer (to ensure uniform sharding).
             """
             if self.ddp_config.use_distributed_optimizer:
-                # Workaround for TE bug causing cuBLAS to pick an incompatible algorithm.
```
Comment (deepakn94) [inline megatron/core/distributed/param_and_grad_buffer.py:835]: Don't delete this.

File: megatron/core/distributed/param_and_grad_buffer.py [hunk 2]
```diff
@@ -796,7 +868,18 @@ def __init__(
...
+            layout = optimizer_class.compute_param_layout(
+                params, bucket_size, self.data_parallel_world_size, ddp_config
+            )
+        else:
+            layout = _default_param_layout(params, bucket_size)
+        self.param_index_map = layout.param_index_map
+        self.bucket_indices = layout.bucket_indices
+        per_bucket_numel_unpadded = layout.per_bucket_numel_unpadded
 
         def _pad(number_to_be_padded: int, divisor: int) -> int:
```
Comment (deepakn94) [inline megatron/core/distributed/param_and_grad_buffer.py]: Is this method used at all?

**Review by kvareddy — APPROVED** [review]
LGTM.

**Review by jiemingz — APPROVED** [review]

File: megatron/core/distributed/distributed_data_parallel.py [hunk 3]
```diff
@@ -290,20 +194,111 @@ def _allocate_buffers_for_parameters(
...
                 expert_gradient_scaling_factor = 1.0 / data_parallel_world_size
 
-        # Allocate the param+grad buffers for dense params' grads.
-        self.buffers, self.bucket_groups = _allocate_buffers_for_parameters(
-            dense_params, self.intra_dp_cp_group, gradient_scaling_factor=gradient_scaling_factor
-        )
+        # Allocate buffers for each group.
+        self.buffers = []
+        self.expert_parallel_buffers = []
+        all_buffers = []
```
Comment (deepakn94) [inline megatron/core/distributed/distributed_data_parallel.py]: Why do I need this variable?

File: megatron/core/distributed/distributed_data_parallel.py [hunk 4]
```diff
@@ -290,20 +194,111 @@ def _allocate_buffers_for_parameters(
...
+        pg_collection = ProcessGroupCollection()
```
Comment (skyw) [inline megatron/core/distributed/distributed_data_parallel.py]: Suggestion: Favor setting groups at contructing time, e.g. ProcessGroupCollection(tp=xxx, ...)

File: megatron/core/distributed/param_and_grad_buffer.py [hunk 5]
```diff
@@ -758,6 +758,116 @@ def register_grad_ready(
...
 
+from ..optimizer.param_layout import BufferKey, PerBufferParamLayout
+
+
+def _compute_default_per_buffer_param_layout(
+    params: List[torch.nn.Parameter], bucket_size: Optional[int]
+) -> PerBufferParamLayout:
+    """Compute parameter layout for the non-distributed-optimizer case.
+
+    No padding is applied. Parameters are iterated in reverse order (backprop order)
```
Comment (skyw) [inline megatron/core/distributed/param_and_grad_buffer.py:819]: Nit: explain why iterating in reversed order.

File: megatron/core/distributed/param_and_grad_buffer.py [hunk 6]
```diff
@@ -758,6 +758,116 @@ def register_grad_ready(
...
+        this_numel = param.data.nelement()
+        param_end_index = param_start_index + this_numel
+        param_index_map[param] = (param_start_index, param_end_index, bucket_id)
+        bucket_params.add(param)
+
+        if bucket_size is not None and (param_end_index - bucket_start_index) >= bucket_size:
+            per_bucket_numel_unpadded.append(param_end_index - bucket_start_index)
+            bucket_indices.append((bucket_start_index, param_end_index))
+            bucket_start_index = param_end_index
+            bucket_params = set()
```
Comment (skyw) [inline megatron/core/distributed/param_and_grad_buffer.py:850]: Confusing code. Not clear why it is needed.
If param shouldn't be added when this condition met, bucket_params.add(param) should be moved under this condition.

File: megatron/core/optimizer/param_layout.py [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+    Attributes:
+        param_index_map: Mapping from parameter to (start_index, end_index, bucket_id) in buffer.
+        bucket_indices: List of (start_index, end_index) for each bucket.
+        per_bucket_numel_unpadded: Number of unpadded elements per bucket.
+        param_indices: The index of each param among same-dtype params (using the "fake"
+            high-precision dtype for FP8 params). Needed for loading non-native-fp8
+            checkpoints in native-fp8 mode. Order matches param_index_map iteration order.
+    """
+
+    param_index_map: Dict[torch.nn.Parameter, Tuple[int, int, int]] = field(default_factory=dict)
```
Comment (skyw) [inline megatron/core/optimizer/param_layout.py:74]: Q: Is there any performance concern with large object as dict key? the time spend on hashing.

Comment (deepakn94) [reply megatron/core/optimizer/param_layout.py:74]: PyTorch just looks at the memory location of the underlying weights, so should be cheap (`O(1)`).

**Comment by deepakn94** [comment]
Note to self: move the assertion on [this line](https://github.com/NVIDIA/Megatron-LM/pull/4328/changes#diff-c4ada2b3f7563ba8629e9bb5617ade92aa63194d8490e1fbc17642b26dc66f74R1058) outside the `if self.ddp_config.use_distributed_optimizer` if condition.

File: megatron/core/optimizer/distrib_optimizer.py [hunk 8]
```diff
@@ -470,6 +475,134 @@ def _build_model_and_main_param_groups(
...
+            params: List of parameters to lay out.
+            bucket_size: Approximate number of elements per bucket, or None for single bucket.
+            data_parallel_world_size: Size of the data-parallel group.
+            ddp_config: DistributedDataParallel config object.
+
+        Returns:
+            PerBufferParamLayout with the computed mapping.
+        """
+
+        def _pad_start_of_param(param_start_index: int) -> int:
```
Comment (deepakn94) [inline megatron/core/optimizer/distrib_optimizer.py]: Why do I need these helper methods if they're just calling the imported methods?

File: megatron/core/distributed/param_and_grad_buffer.py [hunk 9]
```diff
@@ -753,6 +754,116 @@ def register_grad_ready(
...
+            bucket_id += 1
+            param_start_index = param_end_index
```
Comment (deepakn94) [inline megatron/core/distributed/param_and_grad_buffer.py]: Duplicated across if and else blocks. Move out of if-else to de-duplicate.

**Review by yaoyu-33 — APPROVED** [review]
